### PR TITLE
fix(router): make remove trailing slash consistent with URL params

### LINKED ISSUE #16069: ng2 path normalization doesn't work if query parameters are present
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Angular's path normalization usually strips trailing slashes from URLs, redirecting /foo/ -> /foo.  This ensures that URLs have a standard format and that mistakenly adding a slash to the end of a URL doesn't result in a page not found error.  

This logic currently does not work if the URL contains a query parameter.  As an example /foo/?bar=baz results in page not found behavior.

**Expected behavior**
I'd expect /foo/?bar=baz to be normalized to /foo?bar=baz.

**Minimal reproduction of the problem with instructions**
1. Download https://github.com/bourey/ngupgrade-example/tree/ng2
2. npm start
3. npm install
4. Visit http://localhost:3000/teams/?foo=bar.  Compare this to the behavior for http://localhost:3000/teams?foo=bar

**Please tell us about your environment:**

* **Angular version:** 4.0.2
* **Browser:** Chrome 
* **Language:** TS

### PR DESCRIPTION
**Pull request by jasonaden**
closes #16069

### CODE REVIEW

File: packages/common/src/location/location.ts [hunk 1]
```diff
@@ -170,7 +170,9 @@ export class Location {
   /**
    * If url has a trailing slash, remove it, otherwise return url as is.
    */
-  public static stripTrailingSlash(url: string): string { return url.replace(/\/$/, ''); }
+  public static stripTrailingSlash(url: string): string {
+    return url.replace(/\/(?:\?.+)?$/, (match, idx, input) => input.slice(idx + 1));
```
Comment (vicb) [inline packages/common/src/location/location.ts]: `path/to?test/?=3` you don't want to strip here

File: packages/common/test/location/location_spec.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import {Injector} from '@angular/core';
+import {ComponentFixture, TestBed, async, getTestBed} from '@angular/core/testing';
+
+import {Location} from '../../src/location/location';
+import {LocationStrategy} from '../../src/location/location_strategy';
+import {MockLocationStrategy} from '../../testing'
+
+export function main() {let baseUrl = '/'; let parentUrl = '/parent'; let childUrl = '/parent/child'
+
+  fdescribe('Location Class', () => {
```
Comment (vicb) [inline packages/common/test/location/location_spec.ts]: remove fdescribe

File: packages/common/test/location/location_spec.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ */
+
+import {Injector} from '@angular/core';
+import {ComponentFixture, TestBed, async, getTestBed} from '@angular/core/testing';
+
+import {Location} from '../../src/location/location';
+import {LocationStrategy} from '../../src/location/location_strategy';
+import {MockLocationStrategy} from '../../testing'
+
+export function main() {let baseUrl = '/'; let parentUrl = '/parent'; let childUrl = '/parent/child'
```
Comment (vicb) [inline packages/common/test/location/location_spec.ts]: missing semi colon at the end
let -> const

File: packages/common/test/location/location_spec.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+  fdescribe('Location Class', () => {
+  it('should normalize strip a trailing slash', () => {
+    let input = parentUrl + '/';
```
Comment (vicb) [inline packages/common/test/location/location_spec.ts]: let -> const

File: packages/common/test/location/location_spec.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    expect(Location.stripTrailingSlash(input)).toBe(input);
+  })
+
+  it('should not remove slashes after a pound sign', () => {
+    let input = parentUrl + '#test/?=3';
+    expect(Location.stripTrailingSlash(input)).toBe(input);
+  })
+  });
+
+  describe('Location', () => {
```
Comment (vicb) [inline packages/common/test/location/location_spec.ts]: is this test useful ?

File: packages/common/test/location/location_spec.ts (hunk 2 above)
Comment (vicb) [inline packages/common/test/location/location_spec.ts]: nested describe for `strip trailing slash` ?

File: packages/common/src/location/location.ts [hunk 6]
```diff
@@ -170,7 +170,13 @@ export class Location {
   /**
    * If url has a trailing slash, remove it, otherwise return url as is.
    */
-  public static stripTrailingSlash(url: string): string { return url.replace(/\/$/, ''); }
+  public static stripTrailingSlash(url: string): string {
+    const hasSeparator = url.match(/#|\?/);
```
Comment (vicb) [inline packages/common/src/location/location.ts]: hasSeparator -> match ?

File: packages/common/src/location/location.ts [hunk 7]
```diff
@@ -170,7 +170,13 @@ export class Location {
...
+    const pathEndIdx = hasSeparator ? hasSeparator.index : url.length;
+    let urlPart = url.slice(0, pathEndIdx);
+    urlPart = urlPart.replace(/\/(?:\?.+)?$/, (match, idx, input) => input.slice(idx + 1));
+    return urlPart + url.slice(pathEndIdx);
```
Comment (vicb) [inline packages/common/src/location/location.ts]:
```
  const match = url.match(/#|\?/);
  const pathEndIdx = match ? match.index - 1 : url.length - 1;
  const strippedPathEndIx = url[pathEndIdx] === '/' ? pathEndIdx - 1: pathEndIdx;
  return url.slice(0, strippedPathEndIx + 1) + url.slice(pathEndIdx + 1);
```

File: packages/common/test/location/location_spec.ts (hunk 5 above)
Comment (jasonaden) [inline packages/common/test/location/location_spec.ts]: Removed

File: packages/common/test/location/location_spec.ts (hunk 2 above)
Comment (jasonaden) [inline packages/common/test/location/location_spec.ts]: Done

Comment (jasonaden) [inline packages/common/test/location/location_spec.ts]: Done.

File: packages/common/test/location/location_spec.ts (hunk 3 above)
Comment (jasonaden) [inline packages/common/test/location/location_spec.ts]: done.

File: packages/common/src/location/location.ts (hunk 7 above)
Comment (jasonaden) [inline packages/common/src/location/location.ts]: Went with a slightly different approach as there was a minor issue with the above, but hopefully simplified the logic a little bit.

File: packages/common/test/location/location_spec.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+import {Injector} from '@angular/core';
+import {ComponentFixture, TestBed, async, getTestBed} from '@angular/core/testing';
+
+import {Location} from '../../src/location/location';
+
+export function main() {
+  const baseUrl = '/';
+  const parentUrl = '/parent';
+  const childUrl = '/parent/child';
```
Comment (vicb) [inline packages/common/test/location/location_spec.ts]: base and child are not used

**Review by vicb — APPROVED** [review]

**Pull request closed without merging**